modules/services/repositories/src/main/java/au/com/shiftyjelly/pocketcasts/repositories/playback/CacheWorker.kt

```diff
@@ -0,0 +1,106 @@
+package au.com.shiftyjelly.pocketcasts.repositories.playback
+
+import android.content.Context
+import android.net.Uri
+import androidx.annotation.OptIn
+import androidx.hilt.work.HiltWorker
+import androidx.media3.common.util.UnstableApi
+import androidx.media3.datasource.DataSpec
+import androidx.media3.datasource.cache.CacheDataSource
+import androidx.media3.datasource.cache.CacheWriter
+import androidx.work.Constraints
+import androidx.work.Data
+import androidx.work.NetworkType
+import androidx.work.OneTimeWorkRequest
+import androidx.work.WorkManager
+import androidx.work.Worker
+import androidx.work.WorkerParameters
+import dagger.assisted.Assisted
+import dagger.assisted.AssistedInject
+import timber.log.Timber
+
+@OptIn(UnstableApi::class)
+@HiltWorker
+class CacheWorker @AssistedInject constructor(
+    @Assisted private val context: Context,
+    @Assisted params: WorkerParameters,
+    private val exoPlayerHelper: ExoPlayerHelper,
+) : Worker(context, params) {
+    private var cacheWriter: CacheWriter? = null
+
+    override fun doWork(): Result {
+        try {
+            val url = inputData.getString(URL_KEY)
+            val episodeUuid = inputData.getString(EPISODE_UUID_KEY)
+            val uri = Uri.parse(url)
+
+            val dataSourceFactory = exoPlayerHelper.getDataSourceFactory()
+            var dataSpec = DataSpec(uri)
+            if (episodeUuid != null) {
+                dataSpec = dataSpec.buildUpon().setKey(episodeUuid).build()
+            }
+
+            val cacheDataSourceFactory = CacheDataSource.Factory()
+                .setUpstreamDataSourceFactory(dataSourceFactory)
+                .setFlags(CacheDataSource.FLAG_IGNORE_CACHE_ON_ERROR)
+            exoPlayerHelper.getSimpleCache()?.let {
+                cacheDataSourceFactory.setCache(it)
+            }
+
+            cacheWriter = CacheWriter(
+                cacheDataSourceFactory.createDataSource(),
+                dataSpec,
+                null,
+            ) { requestLength: Long, bytesCached: Long, _: Long ->
+                if (bytesCached == requestLength) {
+                    Timber.d("$TAG: Caching complete for episode id: $episodeUuid")
+                }
+            }
+            cacheWriter?.cache()
+        } catch (exception: Exception) {
+            Timber.e("$TAG: ${exception.message}")
```

Comment:
A couple of small improvements that logging could use:
* Logging only a message from an exception might be a bit unhelpful. In case of issues we probably wan to have information about the line that failed.
* Use explicit `Timber.tag()` instead of passing a tag in a message. This allows to filter logs by tags.
* It might be helpful to log an id of the episode and an URL that failed.

I propose to use something like this.

```kotlin
private val episodeUuid get() = inputData.getString(EPISODE_UUID_KEY)
private val downloadUrl get() = inputData.getString(URL_KEY)

Timber.tag(TAG).e(exception, "Failed to cache episode '$episodeUuid' for url '$downloadUrl')
```

It might be also worth considering adding worker id to all logging messages for further filtering identification.

Reply:
That's a great suggestion! Addressed in [628fcce](https://github.com/Automattic/pocket-casts-android/pull/2380/commits/628fcce0d5baa70954089ae45439a963a94f8bdf).